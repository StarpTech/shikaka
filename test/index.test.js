const fs = require('fs-extra');
const { resolve, relative } = require('path');
const dirTree = require('directory-tree');
const execa = require('execa');

const FIXTURES_DIR = `${__dirname}/fixtures`;
const DEFAULT_SCRIPT = 'node';

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
const join = (arr, delimiter = '') => arr.join(delimiter);
const printTree = (nodes, indentLevel = 0) => {
  const indent = '  '.repeat(indentLevel);
  return join(
    nodes
      .filter((node) => node.name[0] !== '.')
      .map((node) => {
        const isDir = node.type === 'directory';
        return `${indent}${node.name}\n${isDir ? printTree(node.children, indentLevel + 1) : ''}`;
      })
  );
};

describe('fixtures', () => {
  const dirs = fs
    .readdirSync(FIXTURES_DIR)
    .filter((fixturePath) => fs.statSync(resolve(FIXTURES_DIR, fixturePath)).isDirectory());

  it.each(dirs)('build %s with shikaka', async (fixtureDir) => {
    let fixturePath = resolve(FIXTURES_DIR, fixtureDir);

    await sleep(1);

    const entry = relative(process.cwd(), resolve(fixturePath, 'src', 'index.js'));
    const outDir = relative(process.cwd(), `${fixturePath}/dist`);
    const scriptArgs = ['cli.js', entry, '--out-dir', outDir];
    await execa(DEFAULT_SCRIPT, scriptArgs);

    await sleep(1);

    const printedDir = printTree([dirTree(fixturePath)]);

    expect(
      [
        `Used script: ${DEFAULT_SCRIPT} ${scriptArgs.join(' ')}`,
        'Directory tree:',
        printedDir,
      ].join('\n\n')
    ).toMatchSnapshot();

    const dist = resolve(`${fixturePath}/dist`);
    const files = await fs.readdir(resolve(dist));
    expect(files.length).toMatchSnapshot();

    // we don't realy care about the content of a sourcemap
    files
      .filter((file) => !/\.map$/.test(file) && !fs.lstatSync(resolve(dist, file)).isDirectory())
      .forEach((file) => {
        expect(fs.readFileSync(resolve(dist, file)).toString('utf8')).toMatchSnapshot();
      });
  });
});
