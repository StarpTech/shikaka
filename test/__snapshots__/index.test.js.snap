// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixtures build library with shikaka 1`] = `
"Used script: node cli.js test/fixtures/library/src/index.js --out-dir test/fixtures/library/dist

Directory tree:

library
  dist
    Button.js
    Modal.js
    index.css
    index.js
  src
    components
      Button
        index.js
        index.module.css
      Modal
        index.js
        index.module.scss
    index.js
"
`;

exports[`fixtures build library with shikaka 2`] = `4`;

exports[`fixtures build library with shikaka 3`] = `
"var styles = {\\"button\\":\\"Button__button\\",\\"unused\\":\\"Button__unused\\",\\"bar-1\\":\\"Button__bar-1\\",\\"bar-2\\":\\"Button__bar-2\\"};

// eslint-disable-next-line no-unused-vars
function Button() {
  return /*#__PURE__*/React.createElement(\\"div\\", {
    className: styles.button
  });
}

export default Button;
"
`;

exports[`fixtures build library with shikaka 4`] = `
"// eslint-disable-next-line no-unused-vars
function Modal() {}

export default Modal;
"
`;

exports[`fixtures build library with shikaka 5`] = `
".Modal__modal {
  width: auto;
}
.Modal__modal__foo {
  border: 1px solid;
}
.Button__button {
  width: auto;
}
.Button__unused {
  width: auto;
}
.Button__bar-1 .bar .Button__bar-2 {
  color: white;
}
.page {
  padding: 20px;
}"
`;

exports[`fixtures build library with shikaka 6`] = `
"export { default as Foo } from './Button.js';
"
`;
